flips/context-propagation.md

```diff
@@ -0,0 +1,210 @@
+* For channels where there should only be staked node participants, we can register a [topic validator](https://pkg.go.dev/github.com/libp2p/go-libp2p-pubsub#PubSub.RegisterTopicValidator) which validates the sender of each incoming message by extracting their network key from their peer ID and looking it up in the protocol state. If the peer ID is valid, we can take the corresponding Flow ID of the sender and add it to the `Context` which is passed to the downstream engine. Otherwise, we reject the message.
+* We may be able to encapsulate a lot of the boilerplate code involved in handling startup / shutdown of child routines / sub-components into a single `BaseComponent` which implements all the logic of `Run`, `Ready`, and `Done`:
+
+  ```golang
+  type Component interface {
+    ReadyDoneAware
+    Runnable
+  }
+
+  type BaseComponent struct {
+    SubComponents []Component
+    ChildRoutines []func(context.Context)
+    Cancel        context.CancelFunc
+
+    childRoutinesStarted  sync.WaitGroup
+    childRoutinesFinished sync.WaitGroup
+  }
+
+  func (c *BaseComponent) Run(parent context.Context) {
+    ctx, cancel := context.WithCancel(parent)
+    c.Cancel = cancel
+
+    for _, component := range c.SubComponents {
+      component.Run(ctx)
+    }
+
+    for _, routine := range c.ChildRoutines {
+      c.childRoutinesStarted.Add(1)
+      c.childRoutinesFinished.Add(1)
+
+      go func(routine func(context.Context)) {
+        c.childRoutinesStarted.Done()
+        routine(ctx)
+        c.childRoutinesFinished.Done()
+      }(routine)
+    }
+  }
+
+  func (c *BaseComponent) Ready() <-chan struct{} {
+    ready := make(chan struct{})
+
+    go func() {
+      <-lifecycle.AllReady(c.SubComponents...)
+      c.childRoutinesStarted.Wait()
+      close(ready)
+    }()
+
+    return ready
+  }
+
+  func (c *BaseComponent) Done() <-chan struct{} {
+    done := make(chan struct{})
+
+    go func() {
+      <-lifecycle.AllDone(c.SubComponents...)
+      c.childRoutinesFinished.Wait()
+      close(done)
+    }()
+
+    return done
+  }
+  ```
+
+  Components that want to provide these interfaces can use this `BaseComponent` to simplify implementation:
+
+  ```golang
+  type FooComponent struct {
+    *BaseComponent
+  }
+
+  func NewFooComponent(foo fooType, bar barType) *FooComponent {
+    f := &FooComponent{}
+
+    f.ChildRoutines = append(f.ChildRoutines, f.childRoutine, f.childRoutineWithFooParameter(foo))
+    f.SubComponents = append(f.SubComponents, NewBarComponent(bar))
+
+    return f
+  }
+
+  func (f *FooComponent) childRoutine(ctx context.Context) {
+    for {
+      select {
+      case <-ctx.Done():
+        return
+      default:
+        // do work...
```

Comment:
Not sure if this is a reliable way of life-cycle management for Flow. I acknowledge the Golang best practices around contexts that you also elaborated. The problem is with this approach we keep delegating the life-cycle management of sub-components to themselves, which is in contrast to managing the life-cycle by the parent component. 

For example, assume the `// do work` part here is a long-running task. Canceling the context wouldn't kill that task until it returns and the subsequent iteration of `for` starts. However, the parent component called `ctx.Cancel()` assumes all sub-components have been terminated immediately. Right?

The problem is a faulty component that faced deadlock may never get terminated in this approach, and there is no way that the parent component is aware of it.

Reply:
> For example, assume the // do work part here is a long-running task. Canceling the context wouldn't kill that task until it returns and the subsequent iteration of for starts. However, the parent component called ctx.Cancel() assumes all sub-components have been terminated immediately. Right?

No, it doesn't assume that all sub-components have been terminated immediately. That's what the `Done` method is for. You used the channel returned from `Done` to actually wait for shutdown to complete. the cancelling of the context just triggers the shutdown.